build.py

```diff
@@ -0,0 +1,263 @@
+    permissionParams = []
+    permissionParams.append ('chmod')
+    permissionParams.append ('+x')
+    permissionParams.append (str(devKitFolder / 'Support' / 'Tools' / 'OSX' / 'ResConv'))
+    permissionResult = subprocess.call (permissionParams)
+    return permissionResult == 0
+
+def PrepareDirectories(rootFolder, buildFolder, packageRootFolder, devKitFolderList, args, configData, platformName):
+    acVersionList = args.acVersion.split()
+
+    os.chdir (rootFolder)
+
+    if not buildFolder.exists():
+        buildFolder.mkdir(parents=True)
+
+    if args.package:
+        if (packageRootFolder).exists():
+            shutil.rmtree (packageRootFolder)
+
+    # Set APIDevKit directory if local is used, else create new directories
+    if args.devKitPath is not None:
+        devKitFolderList[acVersionList[0]] = pathlib.Path(args.devKitPath)
+    else:
+        # For every ACVersion
+        # Check if APIDevKitLink is provided
+        # Create directory for APIDevKit
+        # Download APIDevKit
+        for version in acVersionList:
+            if version in configData['devKitLinks']:
+
+                devKitFolder = rootFolder / f'APIDevKit-{version}'
+                if not devKitFolder.exists():
+                    devKitFolder.mkdir()
+
+                devKitFolderList[version] = devKitFolder
+                DownloadAndUnzip(configData['devKitLinks'][version], devKitFolder, platformName)
+
+            else:
+                print('APIDevKit download link not provided')
+                return 1
+        
+    return 0
+
+def DownloadAndUnzip (url, dest, platformName):
+    # https://github.com/GRAPHISOFT/archicad-api-devkit/releases/download/<tag>/ <filename>
+    # https://github.com/GRAPHISOFT/archicad-api-devkit/releases/latest/download/ <filename>   - not used, as there isn't a single latest version
+
+    version = url.split('/')[-1]
+    fileName = f'API.Development.Kit.{platformName}.{version}.zip'
+    url = f'{url}/{fileName}'
+    filePath = pathlib.Path(dest, fileName)
+    if filePath.exists():
+        return
+
+    print (f'Downloading {fileName}')
+    urllib.request.urlretrieve (url, filePath)
+
+    print (f'Unzipping {fileName}')
+    if platform.system () == 'Windows':
+        with zipfile.ZipFile (filePath, 'r') as zip:
+            zip.extractall (dest)
+    elif platform.system () == 'Darwin':
+        subprocess.call ([
+            'unzip', '-qq', filePath,
+            '-d', dest
+        ])
+
+def BuildAddOn (rootFolder, buildFolder, devKitFolder, addOnName, platformName, configuration, optionalParams, languageCode=None):
+    buildPath = buildFolder / addOnName
+    if languageCode is not None:
+        buildPath = buildPath / languageCode
+
+    # Add params to configure cmake
+    projGenParams = []
+    projGenParams.append ('cmake')
+    projGenParams.extend (['-B', str(buildPath)])
+
+    if platformName == 'WIN':
+        projGenParams.extend (['-G', 'Visual Studio 16 2019'])

```

Comment:
Don't we need to set the platform toolset to `v142`?

Reply:
The `v142` toolset is the default one used with Visual Studio 2019. The lack of toolset specification did not influence the build process when we tested it.

Reply:
For VS2019 it's true, but if I have VS2022 on my computer or on the GitHub runner, the Add-On won't work anymore. I think it would be more safe to define the platform toolset here.

Reply:
Since we are using windows-2019 image on the GitHub runner, it will always have VS 2019 installed.
But locally, things are different. I added the toolset definition.

Reply:
It doesn't have to be in the config file, it can be easily "calculated" from the Archicad version:
https://github.com/GRAPHISOFT/archicad-addon-cmake#visual-studio-windows

Reply:
> Since we are using windows-2019 image on the GitHub runner, it will always have VS 2019 installed.
Sooner or later the windows-2019 image will be deprecated, so it won't always be true on GitHub runners, either.

Reply:
> It doesn't have to be in the config file, it can be easily "calculated" from the Archicad version: https://github.com/GRAPHISOFT/archicad-addon-cmake#visual-studio-windows

Right now, there are only 2 toolset options, but soon the compiler will be updated, so is setting the toolset conditionally even a good idea? Or should we create a mapping for the versions, that we update manually each time a new version comes out?

Reply:
In that case, I added a simple if statement to set the platform toolset for now. Later on, we can update the script with different conditions, or add a separate mapping/json file